### PR TITLE
Libraries IA: Rewrite stale PR

### PR DESCRIPTION
**Pull request by pjhampton**
## Description of new Instant Answer, or changes

To start with, this IA will introduce **<strike>18</strike> 16 new package managers** 🤡  🎉  I forked the stale PR from a community member and completely rewrote it to include single view, familiar favicons, infobox data, and fuzzy search.

-  alcatraz (Swift)
-  atom,
-  bower
-  cran (R)
-  Pub (Dart)
- dub (D-lang)
- elm
- emacs
- go
- <strike>groovy</strike> (No longer on libraries.io it seems)
- <strike>haxe</strike> (Already exists)
- inqlude (C++/Qt)
- julia
- meteor (node)
- nimble (nim)
- nuget (c#)
- shards (crystal)
- wordpress

## Images

**Tile View**
screen shot 2017-06-14 at 10 12 14 pm

**Julia detail view**
screen shot 2017-06-14 at 9 13 08 pm

**Elm detail view**
screen shot 2017-06-14 at 9 08 13 pm

**Go Lang Package view**
screen shot 2017-06-14 at 9 07 39 pm

## Related Issues and Discussions

Fixes: https://github.com/duckduckgo/zeroclickinfo-spice/issues/3101

## People to notify

@moollaza 
/ cc. @bsstoner

---

Instant Answer Page: https://duck.co/ia/view/libraries

### CODE REVIEW

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

**Comment by pjhampton** [comment]
I had to drop support for Dart, Groovy and Haxe as the api was returning 500 for everything 😖

**Comment by moollaza** [comment]
> I had to drop support for Dart, Groovy and Haxe as the api was returning 500 for everything 😖

That's unfortunate. I'm assuming it was working before? @andrew, would you happen to know if there were any recent issues with the Libraries.io API?

**Comment by pjhampton** [comment]
Yeah, it was working fine last week. Not sure what happened 👍

**Comment by andrew** [comment]
I'm traveling at the moment, please do open issues on the libraries API bugs

**Comment by andrew** [comment]
Looks like we had one of the web servers fall over whilst I was travelling but it didn't get removed from the load balancer so that could have explained the 500 errors you were seeing,

**Comment by pjhampton** [comment]
Ahh cool. Thanks for response @andrew

File: lib/DDG/Spice/Libraries.pm [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    chomp($line);
+    my @aliases = split(/,/, $line);
+    push(@trigger_aliases, @aliases);
+    $alias_hash{$aliases[0]} = \@aliases;
+}
+
+triggers start => @trigger_aliases;
+
+spice from => '(.+)/(.+)';
+spice to => 'https://libraries.io/api/search?platforms=$1&q=$2&api_key=34b827d44ed29b80497b1634c2e7e115';
```
Comment (moollaza) [inline lib/DDG/Spice/Libraries.pm]: Let's replace this with an ENV Var. I'll put a key into the production machines 👍

Comment (moollaza) [inline lib/DDG/Spice/Libraries.pm]: `DDG_SPICE_LIBRARIESIO_APIKEY`

File: share/spice/libraries/libraries.js [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+        
+        var text_template =  {
+            group: 'text',
+            options: {
+                content: Spice.libraries.content,
+                moreAt: true
+            }
+        }
+
+        var template, meta, signal;
```
Comment (moollaza) [inline share/spice/libraries/libraries.js]: I would initialize template and signal to `tile_template` and `low`. Then you only need to set it for special cases.

File: share/spice/libraries/libraries.js [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+        var pkgname = api_result[0].name;
+
+
+        meta = {
+            search_term: search_items,
+            total: api_result.length,   
+        }
+
+        // determine which template to show and high/low signal
+        if(api_result.length <= 1) {
```
Comment (moollaza) [inline share/spice/libraries/libraries.js]: Shouldn't this be `length === 1`? If length is 0, it would have returned above

Comment (moollaza) [inline share/spice/libraries/libraries.js]: I would write this as `if ( total === 1 ) {`

File: share/spice/libraries/libraries.js [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+            template = text_template;
+            signal = "high";
+        } else if(pkgname.toLowerCase() == query) {
```
Comment (moollaza) [inline share/spice/libraries/libraries.js]: I think would should combine these `else-ifs`? This block looks a little redundant

File: share/spice/libraries/libraries.js [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+            templates: template,
+            sort_fields: {
+                rank: function(a, b) {
+                    return a.rank > b.rank ? -1 : 1;
+                }
+            },
+            sort_default: 'rank',
+            normalize: function(item) {
+
+                console.log(item.name);
```
Comment (moollaza) [inline share/spice/libraries/libraries.js]: whoops, should be removed

File: share/spice/libraries/libraries.js (hunk 4 above)
Comment (moollaza) [inline share/spice/libraries/libraries.js]:
```js
if(pkgname.toLowerCase() === query
    || ( ... )
    || ( ... )) {
              template = "text_template";
              signal = "high";
}

if (total > 1) {
 api_result = api_result.shift();
}
```

File: share/spice/libraries/libraries.js [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+
+        // determine which template to show and high/low signal
+        if(api_result.length <= 1) {
+            template = text_template;
+            signal = "high";
+        } else if(pkgname.toLowerCase() == query) {
+            template = text_template;
+            api_result = api_result[0];
+            signal = "high";
+        } else if(api_result[0].platform === "Elm" && pkgname.split("/")[1] == query) {
```
Comment (moollaza) [inline share/spice/libraries/libraries.js]: nit: mixed use of `===` and `==` -- `===` should be fine for both checks 👍

File: share/spice/libraries/libraries.js [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+        } else {
+            template = tile_template;
+            signal = "low";
+            meta.sourceName = 'libraries.io';
+            meta.sourceUrl = 'https://libraries.io/api/search?q=' + search_items;
+        }
+
+        Spice.add({
+            id: 'libraries',
+            name: 'Libraries',
```
Comment (moollaza) [inline share/spice/libraries/libraries.js]: I think `name` should be `Software` like we do elsewhere.

File: share/spice/libraries/libraries.js [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+                    url = item.homepage;
+                } else if(item.package_manager_url != null && item.package_manager_url != "") {
+                    url = item.package_manager_url;
+                } else if(item.repository_url != null && item.homepage != item.repository_url) {
+                    url = item.repository_url;
+                } else {
+                    url = 'https://libraries.io/api/search?q=' + search_items;
+                }
+
+                if (item.name && item.description) {
```
Comment (moollaza) [inline share/spice/libraries/libraries.js]: Should we bail earlier if these are not true?

i.e. at the top `if (!item.name && item.description) { return }`

Alternatively, you could set a relevancy block to require these to be defined. You'd need to set the relevancy block depending on whether we're doing a list or detail view

File: share/spice/libraries/libraries.js [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+    }
+
+    // sets up the install script for text view
+    // eg. elm-package install elm-lang/html
+    function getInstallScript(pkg_manager, pkg_name, pkg_version) {
+        var index = pkg_manager.toLowerCase();
+        var text = install_scripts[index] || false;
+
+        if(index === "nuget" || index === "pub") {
+            text = text.replace('{{package}}', pkg_name);
```
Comment (moollaza) [inline share/spice/libraries/libraries.js]: nit: you can chain the replaces here: `text = text.replace(...).replace(...)`

**Comment by pjhampton** [comment]
I've updated this to prevent the overtriggering.

**Comment by moollaza** [comment]
@pjhampton Thanks! This LGTM :+1:

**Comment by moollaza** [comment]
@pjhampton Thanks! This LGTM :+1:

**Pull request merged**